test/performance/scale_from_zero_test.go

```diff
@@ -0,0 +1,198 @@
+	names.Route = test.AppendRandomString(routeName, logger)
+
+	if err := test.CreateConfiguration(logger, clients, names, imagePath, options); err != nil {
+		return test.ResourceNames{}, err
+	}
+	err := test.CreateRoute(logger, clients, names)
+	return names, err
+}
+
+func isDeploymentScaledToZero() func(d *v1beta1.Deployment) (bool, error) {
+	return func(d *v1beta1.Deployment) (bool, error) {
+		return d.Status.ReadyReplicas == 0, nil
+	}
+}
+
+func setup(t *testing.T, clients *test.Clients, name string) *testContext {
+	logger := logging.GetContextLogger("TestScaleFromZero (" + name + ")")
+	imagePath := test.ImagePath("helloworld")
+	names, err := createRouteAndConfig(clients, logger, imagePath, &test.Options{})
+	if err != nil {
+		t.Fatalf("Failed to create Route and Configuration: %v", err)
+	}
+	test.CleanupOnInterrupt(func() { tearDown(clients, names, logger) }, logger)
+
+	logger.Info("Waiting for route to be marked ready.")
+	if err := test.WaitForRouteState(clients.ServingClient, names.Route, test.IsRouteReady, "RouteIsReady"); err != nil {
+		t.Fatalf("The Route %s was not marked as Ready to serve traffic: %v", names.Route, err)
+	}
+
+	config, err := clients.ServingClient.Configs.Get(names.Config, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Configuration %s was not updated with the new revision: %v", names.Config, err)
+	}
+	deploymentName := config.Status.LatestCreatedRevisionName + "-deployment"
+
+	route, err := clients.ServingClient.Routes.Get(names.Route, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Error fetching Route %s: %v", names.Route, err)
+	}
+	domain := route.Status.Domain
+
+	return &testContext{
+		deploymentName: deploymentName,
+		logger:         logger,
+		names:          names,
+		domain:         domain,
+	}
+}
+
+func runScaleFromZero(t *testing.T, clients *test.Clients, name string) time.Duration {
+	ctx := setup(t, clients, name)
+	defer tearDown(clients, ctx.names, ctx.logger)
+
+	ctx.logger.Info("Waiting for deployment to scale to zero.")
+	if err := test.WaitForDeploymentState(
+		clients.KubeClient,
+		ctx.deploymentName,
+		isDeploymentScaledToZero(),
+		"DeploymentScaledToZero",
+		2*time.Minute); err != nil {
+		t.Fatalf("Failed waiting for deployment to scale to zero: %v", err)
+	}
+
+	start := time.Now()
+	ctx.logger.Info("Waiting for endpoint to serve request")
+	if _, err := pkgTest.WaitForEndpointState(
+		clients.KubeClient,
+		ctx.logger,
+		ctx.domain,
+		pkgTest.Retrying(pkgTest.MatchesBody(helloWorldExpectedOutput), http.StatusNotFound),
+		"HelloWorldServesText",
+		test.ServingFlags.ResolvableDomain); err != nil {
+		t.Fatalf("The endpoint for Route %s at domain %s didn't serve the expected text \"%s\": %v", ctx.names.Route, ctx.domain, helloWorldExpectedOutput, err)
+	}
+
+	ctx.logger.Info("Request completed")
+	return time.Now().Sub(start)
+}
+
+func parallelScaleFromZero(t *testing.T, count int) []time.Duration {
+	durations := make([]time.Duration, count)
```

Comment:
should we probably use a channel to collect the durations? it could race here.

Reply:
I dont follow where the race could be here - each goroutine writes to a unique index in this array.

Reply:
Indeed there shouldn't be a race here. Should be fine as is.